mindmeld/text_preparation/tokenizers.py

```diff
@@ -0,0 +1,258 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2015 Cisco Systems, Inc. and others.  All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module contains Tokenizers."""
+
+from abc import ABC, abstractmethod
+import logging
+import unicodedata
+
+from .spacy_model_factory import SpacyModelFactory
+from ..constants import UNICODE_NON_LATIN_CATEGORY, UNICODE_SPACE_CATEGORY
+
+logger = logging.getLogger(__name__)
+
+
+class Tokenizer(ABC):
+    """Abstract Tokenizer Base Class."""
+
+    def __init__(self):
+        """Creates a tokenizer instance."""
+        pass
+
+    @abstractmethod
+    def tokenize(self, text):
+        """
+        Args:
+            text (str): Input text.
+        Returns:
+            tokens (List[str]): List of tokens.
+        """
+        raise NotImplementedError("Subclasses must implement this method")
+
+
+class NoOpTokenizer(Tokenizer):
+    """A No-Ops tokenizer."""
+
+    def __init__(self):
+        """Initialize the NoOpTokenizer."""
+        pass
+
+    def tokenize(self, text):
+        """
+        Args:
+            text (str): Input text.
+        Returns:
+            tokens (List[str]): List of tokens.
+        """
+        return [text]
+
+
+class LetterTokenizer(Tokenizer):
+    """A Tokenizer that splits text at the character level."""
+
+    def __init__(self):
+        """Initializes the LetterTokenizer."""
+        pass
+
+    def tokenize(self, text):
+        """
+        Identify tokens in text and create normalized tokens that contain the text and start index.
+        Args:
+            text (str): The text to normalize
+        Returns:
+            tokens (List[Dict]): List of tokenized tokens which a represented as dictionaries.
+                Keys include "start" (token starting index), "end" (token ending index), and
+                "text" (token text). For example: [{"start": 0, "text":"hello", "end":4}]
+        """
+        actions_by_char = LetterTokenizer.get_actions_by_char(text)
+        token_num_by_char = LetterTokenizer.get_token_num_by_char(actions_by_char)
+        return LetterTokenizer.create_tokens(text, token_num_by_char)
+
+    @staticmethod
+    def get_actions_by_char(text):
```

Comment:
I think `get_actions_by_char ` and `get_token_num_by_char ` can be composed into one function that returns the `token_num_by_char ` representation. I think the action representation doesnt provide further information ie it has a direct map to `token_num_by_char`

Reply:
Makes sense, went ahead and combined these functions into one. This simplified the code and removed the need to maintain "actions" at all. So "actions" have been removed.